sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Blob/ContainerRegistryBlobClient.cs

```diff
@@ -0,0 +1,503 @@
+        {
+            Argument.AssertNotNull(stream, "stream");
+
+            using DiagnosticScope scope = _clientDiagnostics.CreateScope($"{nameof(ContainerRegistryBlobClient)}.{nameof(UploadBlob)}");
+            scope.Start();
+            try
+            {
+                string digest = OciBlobDescriptor.ComputeDigest(stream);
+
+                ResponseWithHeaders<ContainerRegistryBlobStartUploadHeaders> startUploadResult =
+                    await _blobRestClient.StartUploadAsync(_repositoryName, cancellationToken).ConfigureAwait(false);
+
+                ResponseWithHeaders<ContainerRegistryBlobUploadChunkHeaders> uploadChunkResult =
+                    await _blobRestClient.UploadChunkAsync(startUploadResult.Headers.Location, stream, cancellationToken).ConfigureAwait(false);
+
+                ResponseWithHeaders<ContainerRegistryBlobCompleteUploadHeaders> completeUploadResult =
+                    await _blobRestClient.CompleteUploadAsync(digest, uploadChunkResult.Headers.Location, null, cancellationToken).ConfigureAwait(false);
+
+                return Response.FromValue(new UploadBlobResult(completeUploadResult.Headers.DockerContentDigest), completeUploadResult.GetRawResponse());
+            }
+            catch (Exception e)
+            {
+                scope.Failed(e);
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Download the manifest for an OCI Artifact.
+        /// </summary>
+        /// <param name="digest">The digest of the manifest to download.</param>
+        /// <param name="cancellationToken"> The cancellation token to use. </param>
+        /// <returns></returns>
+        public virtual Response<DownloadManifestResult> DownloadManifest(string digest, CancellationToken cancellationToken = default)
+        {
+            Argument.AssertNotNullOrEmpty(digest, "digest");
+
+            using DiagnosticScope scope = _clientDiagnostics.CreateScope($"{nameof(ContainerRegistryBlobClient)}.{nameof(DownloadManifest)}");
+            scope.Start();
+            try
+            {
+                Response<ManifestWrapper> response = _restClient.GetManifest(_repositoryName, digest, ManifestMediaType.OciManifest.ToString(), cancellationToken);
+                Response rawResponse = response.GetRawResponse();
+                ManifestMediaType mediaType = rawResponse.Headers.ContentType;
+
+                if (!rawResponse.Headers.TryGetValue("Docker-Content-Digest", out string registryDigest))
+                {
+                    throw _clientDiagnostics.CreateRequestFailedException(rawResponse, "Response did not contain \"Docker-Content-Digest\" header.");
+                }
+
+                if (!ValidateDigest(rawResponse.ContentStream, registryDigest))
+                {
+                    throw _clientDiagnostics.CreateRequestFailedException(rawResponse, "The manifest's digest according to the registry does not match the locally computed digest value.");
```

Comment:
```suggestion
                    throw _clientDiagnostics.CreateRequestFailedException(rawResponse, "The manifest's digest, according to the registry, does not match the locally computed digest value.");
```

perhaps print the two values?